### ISSUE TITLE
RFC: Skip Improvements

### ISSUE THREAD

**Issue by jquick** [inspec/inspec#3158]
### Current state

"Skip" is currently used to identify cases where a control cannot run:

1. The user doesn't want it to run when inheriting the profile.
    Example: Use your colleague's AWS profile but skip all controls that test S3.
2. The control isn't run when certain conditions are met.
    For example don't test MySQL if it isn't installed.
3. A control is skipped if it uses resources that are not supported on the target system.
    For example looking for a "package" when testing an API.

### Direction

Q: What can skip communicate?
A: It is a way to tell in reporting that a test wasn't run for an accepted reason.

Skip means: we don't execute a test we intended to run and let the user know we had to skip it.

Skip doesn't require the user to take any action. We let you know that the test was included, it was considered, but it was no executed for an intended reason. In this sense it's different from passed/failed results (indicating status) and failures (where a test that was run failed to execute).

### Design

To tackle skip, we'd establish:

1. A core use-case for skip is `only_if` in controls or files. Skip tests that don't match this condition and report that they were skipped. (same behavior as before)

2. A non-case are inherited profiles that aren't supported. We see this with combined profile layers. Until now, skip has served the value of outlining tests that were not run because the profile isn't supported. We don't want to use skip for this condition and instead hand it over to the profile inheritance and layering mechanism (and separate reporting). The user knows these tests were skipped, without reporting all of them. (new behavior: don't report these skipped tests)

3. A non-case is skipping controls whose resources are not supported. This is a change from previous convention, which turns some controls from being skipped into failing because they aren't supported. Fail explicitly with the condition that wasn't met. For example: instead of reporting `Config.Value.X != Expectation` we report that `Config is not supported on this system`.

4. A non-case are accepted tests or other exceptions. We already covered `only_if` exceptions and inheritance-based exceptions in 1+2. For acceptance, we want the test to be executed, i.e. this is a different domain than skip, since it covers tests that were executed and only require different reporting. (same behavior, will be covered in a separate RFC) (also see https://github.com/inspec/inspec/issues/3017)

Implicit resource-skips would now be turned into explicit user-defined skips (`only_if`, reported per control) or profile-level skips (`inspec.yml`, reported on the profile). If the user applies a resource on a target that doesn't support it and doesn't limit the profile's supported platforms, it will result in an error i.e. a mandatory user-action is required.

In Summary: turn some implicit skips into explicit actions. Remove other skips that were reported that don't provide value.

### Action Items

1) Remove skip from inherited profiles from showing up in reporting (breaking change)
2) Ensure resources fail that target something that isn't supported
3) Update skip messages on controls (from `only_if`) so its very clear why they were skipped.

**Comment by clintoncwolfe** [comment]
This is a partial cover of some of the problems mentioned on #3017 .

**Comment by james-stocks** [comment]
One comment on this is that `only_if` is occasionally relied upon where a control needs to mix AND and OR logic together.

e.g. if a control requires either 5 things about your iptables config to be true (say A, B, C, D, E) _OR_ iptables to be disabled (F)
There is no way to represent `OR(AND(A, B, C, D, E), F)` in an Inspec control, as there is not yet a `describe.all do` concept.
Currently we can only have `only_if(!F); AND(A, B, C, D, E)`

Edit: I'm not sure this actually overlaps with this RFC, but wanted to mention it as a use case that would hopefully not suffer a regression from whatever changes happen.

**Comment by clintoncwolfe** [comment]
From discussions with users, case 2, 

  The control isn't run when certain conditions are met.
  For example don't test MySQL if it isn't installed.

is dangerous: it should be treated as [an error or failure](https://github.com/inspec/inspec/issues/3174).  Many users set their CI to ignore skips, because of case 3 (platform support).  If those cases are indistinguishable, we'll have users skipping tests and [ignoring critical errors](https://github.com/inspec/inspec/issues/2896) in their test coverage.

**Comment by jerryaldrichiii** [comment]
I'm curious also if this work will make the following more delightful?:

```
raise Inspec::Exceptions::ResourceSkipped, 'Some long message'
```

We have `skip_resource` but it does not raise at the moment. See:

https://github.com/inspec/inspec/blob/20a0b0e025185e41b9c3cd6bb112b1accdad049e/lib/inspec/plugins/resource.rb#L98

**Comment by jquick** [comment]
This is done.
